samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,314 @@
+        Not using create_change_set as it is slow.
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of ECR repos deployed for this companion stack
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        repos: List[ECRRepo] = list()
+        stack = boto3.resource("cloudformation", config=self._boto_config).Stack(self._companion_stack.stack_name)
+        for resource in stack.resource_summaries.all():
+            if resource.resource_type == "AWS::ECR::Repository":
+                repos.append(
+                    ECRRepo(logical_id=resource.logical_resource_id, physical_id=resource.physical_resource_id)
+                )
+        return repos
+
+    def get_unreferenced_repos(self) -> List[ECRRepo]:
+        """
+        List deployed ECR repos that is not referenced by current list of functions
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of deployed ECR repos that is not referenced by current list of functions
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        deployed_repos: List[ECRRepo] = self.list_deployed_repos()
+        current_mapping = self._builder.repo_mapping
+
+        unreferenced_repos: List[ECRRepo] = list()
+        for deployed_repo in deployed_repos:
+            for _, current_repo in current_mapping.items():
+                if current_repo.logical_id == deployed_repo.logical_id:
+                    break
+            else:
+                unreferenced_repos.append(deployed_repo)
+        return unreferenced_repos
+
+    def delete_unreferenced_repos(self) -> None:
+        """
+        Blocking call to delete all deployed ECR repos that are unreferenced by a function
+        If repo does not exist, this will simply skip it.
+        """
+        repos = self.get_unreferenced_repos()
+        for repo in repos:
+            try:
+                self._ecr_client.delete_repository(repositoryName=repo.physical_id, force=True)
+            except self._ecr_client.exceptions.RepositoryNotFoundException:
+                pass
+
+    def sync_repos(self) -> None:
+        """
+        Blocking call to sync companion stack with the following actions
+        Creates the stack if it does not exist, and updates it if it does.
+        Deletes unreferenced repos if they exist.
+        Deletes companion stack if there isn't any repo left.
+        """
+        exists = self.does_companion_stack_exist()
+        has_repo = bool(self.get_repository_mapping())
+        if exists:
+            self.delete_unreferenced_repos()
+            if has_repo:
+                self.update_companion_stack()
+            else:
+                self.delete_companion_stack()
+        elif not exists and has_repo:
```

Comment:
Not sure the `not exists` is necessary here.
Which means you can probably directly `if self.does_companion_stack_exist():`

Reply:
Removed not exists and uses if self.does_companion_stack_exist():